src/compiler/checker.ts

```diff
@@ -29996,13 +29996,14 @@ namespace ts {
             rightType = checkNonNullType(rightType, right);
             // TypeScript 1.0 spec (April 2014): 4.15.5
             // The in operator requires the left operand to be of type Any, the String primitive type, or the Number primitive type,
-            // and the right operand to be of type Any, an object type, or a type parameter type.
+            // and the right operand not to extend a primitive type.
             // The result is always of the Boolean primitive type.
             if (!(allTypesAssignableToKind(leftType, TypeFlags.StringLike | TypeFlags.NumberLike | TypeFlags.ESSymbolLike) ||
                   isTypeAssignableToKind(leftType, TypeFlags.Index | TypeFlags.TemplateLiteral | TypeFlags.StringMapping | TypeFlags.TypeParameter))) {
                 error(left, Diagnostics.The_left_hand_side_of_an_in_expression_must_be_of_type_any_string_number_or_symbol);
             }
-            if (!allTypesAssignableToKind(rightType, TypeFlags.NonPrimitive | TypeFlags.InstantiableNonPrimitive)) {
+            if (!allTypesAssignableToKind(rightType, TypeFlags.NonPrimitive | TypeFlags.InstantiableNonPrimitive) ||
+                rightType.immediateBaseConstraint && allTypesAssignableToKind(rightType.immediateBaseConstraint, TypeFlags.Primitive)) {

```

Comment:
Thank you all for the comments and hints!

I wasn't able to extract the information we need directly from `rightType`s flags. Both `alsoValidHasKey` and `invalidHasKey` generate the same type flag for the right type. I guess this is to be expected though as we're looking at constraints on the type to identify whether this is a valid/invalid use of the `in` operator.

Reply:
I’m not 100% on this, but I believe all you need is

```suggestion
            if (!allTypesAssignableToKind(getApparentType(rightType), TypeFlags.NonPrimitive)) {
```

`getApparentType` will get the (resolved) base constraint if the type is instantiable (the immediate base constraint might _also_ be instantiable), which means the result should never be instantiable, so you no longer have to check for `InstantiableNonPrimitive`. But make sure I’m not making stuff up by running the tests, and someone else on the team should double check my understanding here too.

Reply:
@andrewbranch I tried this out both comparing against `NonPrimitive` and comparing against `NonPrimitive | InstantiableNonPrimitive`, but both approaches lead to many scenarios where don't throw an error even though we should.

For example the following is accepted:
```
var b1: number;
var rb1 = 'x' in b1;
```